ce/api/streamflow/watershed_streams.py

```diff
@@ -0,0 +1,164 @@
+    direction_matrix = VIC_direction_matrix(
+        flow_direction.lat_step, flow_direction.lon_step
+    )
+
+    with Timer() as watershed_time:
+        watershed_xys = build_watershed_streams(
+            flow_direction.lonlat_to_xy(station_lonlat),
+            flow_direction.values,
+            direction_matrix,
+            debug=True,
+        )
+
+    # `watershed_lonlats` must be an ordered collection (not sets) because
+    # a multi line string is an array (python list) of linestrings
+    watershed_lonlats = [[]]
+
+    watershed_lonlats = [
+        [flow_direction.xy_to_lonlat(xy) for xy in stream] for stream in watershed_xys
+    ]
+
+    lines = MultiLineString(watershed_lonlats)
+
+    return {
+        "streams": geojson_feature(lines,),
+        "debug/test": {"watershed_streams": {"time": watershed_time.elapsed,}},
+    }
+
+
+def build_watershed_streams(target, routing, direction_map, debug=False):
+    """
+    Returns a set of frozensets, each frozenset representing a path of cells that
+    drain into `target`.
+
+    :param target: An xy index representing the cell of interest.
+    :param routing: A numpy array representing water flow using VIC direction
+        codes (0 - 9). These codes may be integers or floats.
+    :param direction_map: Maps VIC direction codes into offsets in cell indices.
+        Necessary because, depending on whether lon and lat dimensions
+        increase or decrease with increasing index, a move north or east is
+        represented by an offset of +1 or -1, respectively.
+        TODO: Compute this internally?
+    :param debug: Boolean indicating whether this function should compute
+        and return debug information.
+    :return: Set of frozensets of cells (streams) that drain into `target`.
+
+    Notes:
+
+    - In this function, a cell is represented by an (x, y) index pair.
+
+    - Routing graphs can and in practice do contain cycles. Variable `visited`
+    is used to determine whether a cell has already been visited during the
+    traversal of the routing graph, i.e., whether we are cycling, and if so
+    not to repeat that subgraph.
+    """
+    visited = set()
+    counter = 0
+    connection = [[]]
+
+    def upstream(cell):
+        """Return list of lists, each list being a stream of cells upstream of
+        the inputted cell"""
+        nonlocal visited
+        nonlocal counter
+        nonlocal connection
+
+        visited |= {cell}
+        connection[counter].append(cell)
+        for neighbour in neighbours(cell):
+            if neighbour not in visited and is_upstream(
+                neighbour, cell, routing, direction_map
+            ):
+                upstream(neighbour)
+                connection.append([])
+                counter += 1
+                connection[counter].append(cell)
+        return connection
+
+    connection = upstream(target)
+    return set(frozenset(stream) for stream in connection if len(stream) > 1)
```

Comment:
Why are these turned into frozensets? And not left as just lists or transformed into tuples, the immutable version of lists?

Reply:
this was done for ease of testing, so that the order didn't matter when writing tests. It would probably make sense to just keep it as list of lists the entire time and write the tests to be exact, as they are for the ```worker``` class. I can switch these if that is what you think is best